src/node.c

```diff
@@ -150,6 +155,15 @@ cmark_node_type cmark_node_get_type(cmark_node *node) {
   }
 }
 
+int cmark_node_set_type(cmark_node * node, cmark_node_type type) {
+  if (type == node->type)
+    return true;
+
+  free_node_as(node);
+  node->type = type;
+  return true;
+}
```

Comment:
I can see why extensions need this function, but in its current form, usage should be restricted to extension authors who know what they're doing. A safe public function should

- check whether the new node type is valid for the node's parent.
- initialize the node to its default state (unless the content can be reused).

Maybe the parser could first assign an internal node type like UNKNOWN_BLOCK which can later be changed to a known block type?